#18: Add utilities for assisting with common migration logic

cpointe-ibllanos opened the pull request:
Add: file utils, common utils for use in migration logic by users
 Working example for using utils to write a migration
 Update docs to mention utils

## BDDs
* Scenario: I can leverage migration utilities from Baton in my custom migration implementations
** *Given* I have created a new baton migration that imports the baton utils class
** *And* the util methods are used or extended inside the custom migration
** *When* The migration runs
** *Then* The the migration should finish successfully

 * Scenario: I want to easily replace a string in a file using a regex expression
 ** *Given* I have a file containing the string "123abc456"
 ** *When* I use the regex "(?<=123).*(?=456)" to substitute "def"
 ** *Then* the file should now contain the string "123def456"

 * Scenario: I want to easily replace a substring of a string in a file using a regex expression
 ** *Given* I have a file containing the string "123abc456"
 ** *When* I use the regex "123.*456" to substitute "abc" with "def"
 ** *Then* the file should now contain the string "123def456"

 * Scenario: I want to easily test if a string is present in a file using a regex expression
 ** *Given* I have a file containing the string "123abc456"
 ** *When* I use the regex "123.*456" to search for matches in a file
 ** *Then* the match result should be "true"

 * Scenario: I want to easily be able to pull out specific substrings from a string using regex capture groups
 ** *Given* I have a string containing "123abc456"
 ** *When* I use the regex "123(.*)456" to retrieve capture groups
 ** *Then* the size of the retrieved capture groups should be "1"
 ** *And* the capture groups should include
|abc|

 * Scenario: I want to easily be able to pull out specific strings from a file using regex capture groups
 ** *Given* I have a file containing the string "123abc456"
 ** *When* I use the regex "(123).*(456)" to retrieve capture groups
 ** *Then* the size of the retrieved capture groups should be "2"
 ** *And* the capture groups should include
|123|
|456|

## Test Steps
1. Pull down and open Baton
* `git clone git@github.org:TechnologyBrewery/baton`
2. Confirm the Baton bootstrap build is successful and tests pass
* `mvn clean install -Pbootstrap -Dmaven.build.cache.enabled=false`
3. Confirm the Baton example projects build successfully
* `mvn clean install -Pdefault -Dmaven.build.cache.enabled=false`

d-ryan-ashcraft commented on baton-maven-plugin/src/main/java/org/technologybrewery/baton/util/CommonUtils.java:
```diff
@@ -0,0 +1,16 @@
+package org.technologybrewery.baton.util;
+import com.vdurmont.semver4j.Semver;
+
+public class CommonUtils {
```
S: all static classes like this should be `final` with a private constructor to prevent instantiation.  A tool like SonarLint can be useful  for stuff like this.

d-ryan-ashcraft commented on baton-maven-plugin/src/main/java/org/technologybrewery/baton/util/CommonUtils.java:
```diff
@@ -0,0 +1,16 @@
+package org.technologybrewery.baton.util;
+import com.vdurmont.semver4j.Semver;
+
```
A: let's add a comment about the intent - I assume here this is meant for common functions used while _performing_ migrations?

cpointe-ibllanos replied:
Correct. Addressed

d-ryan-ashcraft commented on baton-maven-plugin/src/main/java/org/technologybrewery/baton/util/FileUtils.java:
```diff
@@ -0,0 +1,204 @@
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public final class FileUtils {
```
S: same comments in general as above:
* Add a comment for all classes
* Add a private constructor here too

d-ryan-ashcraft commented on baton-maven-plugin/src/main/java/org/technologybrewery/baton/util/FileUtils.java:
```diff
@@ -0,0 +1,204 @@
+public final class FileUtils {
+
+    /**
+     *
```
S: appreciate the input/output comments, but let's put on for the method too so it's fully javadoc conformant.

cpointe-ibllanos committed:
#18
 Add: file utils, common utils for use in migration logic by users
 Working example for using utils to write a migration
 Update docs to mention utils

JeffreyRoss commented on examples/example-migration-configuration/src/main/java/org/technologybrewery/baton/ReplaceRegexMatchUsingUtilsMigration.java:
```diff
@@ -0,0 +1,41 @@
+            logger.info("Detecting migrations for {} to {}", JAVA_FILE_REGEX, NEW_JAVA_FILE_NAME);
+            shouldExecuteMigration = FileUtils.hasRegExMatch(JAVA_FILE_REGEX, file);
+        } catch (Exception e) {
+            logger.error("Caught exception checking Java files for migration legibility", e);
```
S: I'm still a newbie to Baton but it appears there's a BatonException class, would it be better for consistency to incorporate that special exception logic to handle your exceptions?

cpointe-ibllanos replied:
Yep, good catch. For consistency, this now catches an IOException, logs an error, and throws the exception as a BatonException.

d-ryan-ashcraft commented on baton-maven-plugin/src/main/java/org/technologybrewery/baton/util/CommonUtils.java:
```diff
@@ -0,0 +1,24 @@
+     * @param version2
+     * @return isLessThanVersion - if true, version1 is less than version2.
+     */
+    private CommonUtils(){
```
A: seems like the private constructor got injected before the method.  There is some extra whitespace above too

d-ryan-ashcraft commented on baton-maven-plugin/src/main/java/org/technologybrewery/baton/util/FileUtils.java:
```diff
@@ -0,0 +1,213 @@
+ * Common file-related logic used when performing a migration
+ */
+public final class FileUtils {
+    private static final String SPACE = " ";
```
A: FWIW, I normally just go w/ `StringUtils.SPACE` if I need this.

d-ryan-ashcraft commented on baton-maven-plugin/src/main/java/org/technologybrewery/baton/util/FileUtils.java:
```diff
@@ -0,0 +1,213 @@
+    }
+
+    /**
+     * Evaluates a file against a literal string and replaces each match with a specified replacement
```
A: very minor, but a period at the end of the first sentence in Javadoc really improves functionality (whole sentence shows up in the summary instead the first ~40 characters)

cpointe-ibllanos replied:
The more you know!

d-ryan-ashcraft reviewed (approved):
Thanks for the updates.  A couple minor items I'd recommend cleaning up before committing.
